runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/WindmillStreamSender.java

```diff
@@ -0,0 +1,157 @@
+    // All streams are memoized/cached since they are expensive to create and some implementations
+    // perform side effects on construction (i.e. sending initial requests to the stream server to
+    // initiate the streaming RPC connection). Stream instances connect/reconnect internally so we
+    // can reuse the same instance through the entire lifecycle of WindmillStreamSender.
+    this.getDataStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createGetDataStream(
+                    stub, streamingEngineThrottleTimers.getDataThrottleTimer()));
+    this.commitWorkStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createCommitWorkStream(
+                    stub, streamingEngineThrottleTimers.commitWorkThrottleTimer()));
+    this.getWorkStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createDirectGetWorkStream(
+                    stub,
+                    withRequestBudget(getWorkRequest, getWorkBudget.get()),
+                    streamingEngineThrottleTimers.getWorkThrottleTimer(),
+                    getDataStream,
+                    commitWorkStream,
+                    processWorkItem));
+  }
+
+  public static WindmillStreamSender create(
+      CloudWindmillServiceV1Alpha1Stub stub,
+      GetWorkRequest getWorkRequest,
+      GetWorkBudget getWorkBudget,
+      GrpcWindmillStreamFactory streamingEngineStreamFactory,
+      ProcessWorkItem processWorkItem) {
+    return new WindmillStreamSender(
+        stub,
+        getWorkRequest,
+        new AtomicReference<>(getWorkBudget),
+        streamingEngineStreamFactory,
+        processWorkItem);
+  }
+
+  private static GetWorkRequest withRequestBudget(GetWorkRequest request, GetWorkBudget budget) {
+    return request.toBuilder().setMaxItems(budget.items()).setMaxBytes(budget.bytes()).build();
+  }
+
+  @SuppressWarnings("ReturnValueIgnored")
+  public void startStreams() {
+    Preconditions.checkState(
+        !getWorkBudget.get().equals(GetWorkBudget.noBudget()), "Cannot GetWork with no budget.");
+    getWorkStream.get();
+    getDataStream.get();
+    commitWorkStream.get();
+    // *stream.get() is all memoized in a threadsafe manner.
+    started.compareAndSet(false, true);
+  }
+
+  public void closeAllStreams() {
+    // Supplier<Stream>.get() starts the stream which is an expensive operation as it initiates the
+    // streaming RPCs by possibly making calls over the network. Do not close the streams unless
+    // they have already been started.
+    if (started.get()) {
+      getWorkStream.get().close();
+      getDataStream.get().close();
+      commitWorkStream.get().close();
+    }
+  }
+
+  public synchronized void adjustBudget(long itemsDelta, long bytesDelta) {
+    getWorkBudget.set(getWorkBudget.get().apply(itemsDelta, bytesDelta));
+    getWorkStream.get().adjustBudget(itemsDelta, bytesDelta);
```

Comment:
only do this if started? or enforce that started was called?
this seems like it will create getworkstream with possibly uncached getData/commitWork streams otherwise

Reply:
good point added